File: QtClient/QtClimaClient/Network/GenericServices/HeaterControllerService.cpp
#include "HeaterControllerService.h"

#include <Network/GenericServices/Messages/DefaultRequest.h>
#include <Network/GenericServices/Messages/HeaterParamsListResponse.h>
#include <Network/GenericServices/Messages/HeaterParamsRequest.h>
#include <Network/GenericServices/Messages/HeaterStateListResponse.h>
#include <Network/GenericServices/Messages/HeaterStateRequest.h>
#include <Network/GenericServices/Messages/HeaterStateResponse.h>

HeaterControllerService::HeaterControllerService(QObject *parent) : INetworkService(parent)
{

}

void HeaterControllerService::GetHeaterStates()
{
    DefaultRequest *request = new DefaultRequest();
    request->jsonrpc = "0.1a";
    request->service = "HeaterControllerService";
    request->method = "GetStateList";

    emit SendRequest(request);
}

void HeaterControllerService::GetHeaterParams()
{
    DefaultRequest *request = new DefaultRequest();
    request->jsonrpc = "0.1a";
    request->service = "HeaterControllerService";
    request->method = "GetHeaterParamsList";

    emit SendRequest(request);
}

void HeaterControllerService::UpdateHeaterParams(QList<HeaterParams> heaterParams)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "HeaterControllerService";
    request->method = "UpdateHeaterParamsList";

    HeaterParamsRequest par;
    par.ParamsList = heaterParams;

    request->params = par.toJsonString();

    emit SendRequest(request);
}

void HeaterControllerService::UpdateHeaterState(const QString &key, HeaterState state)
{
    NetworkRequest *request = new NetworkRequest();
    request->jsonrpc = "0.1a";
    request->service = "HeaterControllerService";
    request->method = "UpdateHeaterState";
    HeaterStateRequest st;

    st.Key = key;
    st.State = state;
    request->params = st.toJsonString();

    emit SendRequest(request);
}

void HeaterControllerService::ProcessReply(const NetworkResponse &reply)
{
    if(reply.service == "HeaterControllerService")
    {
        if(reply.method == "GetStateList")
        {
            HeaterStateListResponse resp;
            resp.fromJson(reply.result.toUtf8());

            emit HeaterStateListReceived(resp.SetPoint, resp.Front, resp.Rear, resp.States);
        }
        else if(reply.method == "SetHeaterState")
        {
            HeaterStateResponse resp;
            resp.fromJson(reply.result.toUtf8());
            emit HeaterStateUpdated(resp.State);
        }
        else if(reply.method == "GetHeaterParamsList")
        {
            HeaterParamsListResponse resp;
            resp.fromJson(reply.result.toUtf8());
            emit HeaterParamsListReceived(resp.ParamsList);
        }
        else if(reply.method == "UpdateHeaterParamsList")
        {
            HeaterParamsListResponse resp;
            resp.fromJson(reply.result.toUtf8());
            emit HeaterParamsUpdated(resp.ParamsList);
        }
    }
}

File: ClimaDaemon/CoreImplementations/Clima.Core.Controllers/Network/Services/HeaterControllerService.cs
using System.Linq;
using Clima.Basics.Services.Communication;
using Clima.Core.Controllers.Heater;
using Clima.Core.Controllers.Network.Messages;
using Clima.Core.Network.Messages;

namespace Clima.Core.Controllers.Network.Services
{
    public class HeaterControllerService:INetworkService
    {
        private readonly IHeaterController _heaterController;

        public HeaterControllerService(IHeaterController heaterController)
        {
            _heaterController = heaterController;
        }

        [ServiceMethod]
        public HeaterStateResponse GetHeaterState(HeaterStateRequest request)
        {
            return new HeaterStateResponse()
            {
                State = _heaterController.GetHeaterState(request.Key)
            };
        }
        
        [ServiceMethod]
        public HeaterStateResponse UpdateHeaterState(HeaterStateRequest request)
        {
            _heaterController.UpdateHeaterState(request.Key, request.State);
            return new HeaterStateResponse()
            {
                State = _heaterController.GetHeaterState(request.Key)
            };
        }

        [ServiceMethod]
        public HeaterStateListResponse GetStateList(DefaultRequest request)
        {
            var list = new HeaterStateListResponse();
            list.States = _heaterController.States.Values.ToList();
            list.SetPoint = _heaterController.SetPoint;
            list.Front = ClimaContext.Current.Sensors.FrontTemperature;
            list.Rear = ClimaContext.Current.Sensors.RearTemperature;
            return list;
        }

        [ServiceMethod]
        public HeaterParamsListResponse UpdateHeaterParamsList(HeaterParamsRequest request)
        {
            return new HeaterParamsListResponse() {ParamsList = _heaterController.UpdateHeaterParams(request.ParamsList)};
        }

        [ServiceMethod]
        public HeaterParamsListResponse GetHeaterParamsList(DefaultRequest request)
        {
            return new HeaterParamsListResponse() {ParamsList = _heaterController.Params.Values.ToList()};
        }
    }
}